### PR TITLE
bpo-44490: Add __parameters__ and __getitem__ to types.Union

### PR DESCRIPTION
**Pull request by uriyyo**
https://bugs.python.org/issue44490

### CODE REVIEW

**Comment by uriyyo** [comment]
@Fidget-Spinner Could you please review this PR?

File: Lib/test/test_types.py
Comment (Fidget-Spinner) [inline Lib/test/test_types.py]: I recommend using the unittest assert for better error messages:
```suggestion
        self.assertEqual((float | list[T])[int], float | list[int])
```

File: Misc/NEWS.d/next/Core and Builtins/2021-07-01-11-59-34.bpo-44490.xY80VR.rst
Comment (Fidget-Spinner) [inline Misc/NEWS.d/next/Core and Builtins/2021-07-01-11-59-34.bpo-44490.xY80VR.rst]:
```suggestion
Add ``__parameters__`` attribute and ``__getitem__``
operator to ``types.Union``. Patch provided by Yurii Karabas.
```

File: Objects/unionobject.c
Comment (Fidget-Spinner) [inline Objects/unionobject.c]: I know GenericAlias did it this way, but here's a minor suggestion you can choose to ignore 😅 
```suggestion
union_parameters(PyObject *self, void *Py_UNUSED(unused))
```

File: Include/genericaliasobject.h
Comment (Fidget-Spinner) [inline Include/genericaliasobject.h]: Sorry to trouble, but personally I didn't understand what the ``apply`` here meant. Maybe rename it to ``subs/substitute`` please:
```suggestion
PyAPI_FUNC(PyObject *) _Py_subs_parameters(PyObject *, PyObject *, PyObject *, PyObject *);
```
same for the other call sites below.

**Review by Fidget-Spinner — APPROVED** [review]
@uriyyo thank you for sending this PR so quickly! Excellent work. I only have a few minor readability nits below.

File: Lib/test/test_types.py
Comment (Fidget-Spinner) [inline Lib/test/test_types.py:678]: Oh can you add one more test too please?
Specifically:
```suggestion
self.assertEqual(list[int | list[T]].__parameters__, (T,))
self.assertEqual(list[int | list[T]][str], list[int | list[str]])
```
The idea is to make sure GenericAlias always looks into nested Unions for ``__parameters__``, and nested unions inside GenericAliases work.

**Comment by uriyyo** [comment]
@Fidget-Spinner Thanks for review, I have added tests that you suggested.

**Comment by Fidget-Spinner** [comment]
> @Fidget-Spinner Thanks for review, I have added tests that you suggested.

Thanks for applying the suggestions. This LGTM. I've requested a review from Guido and I'm waiting for him to take a look when he has the time. I can't merge PRs, sorry.

**Comment by Fidget-Spinner** [comment]
@uriyyo can you please rebase this against main branch? I fixed a GC bug in Union in https://github.com/python/cpython/commit/1097384ce964dd63686b1aac706cd0fa764c2dc9, and I think you may have to add ``Py_VISIT(alias->parameters)`` in ``union_traverse`` Thanks.

**Comment by uriyyo** [comment]
@Fidget-Spinner Thanks, fixed

File: Lib/test/test_types.py
Comment (gvanrossum) [inline Lib/test/test_types.py:675]: Add a check showing that having S and T in different parts gets unified? E.g. `list[T] | list[S] --> __parameters__ == (T, S)` and if we substitute something crazy like `[int, T]` it becomes `list[int] | list[T]`.

File: Objects/genericaliasobject.c
Comment (gvanrossum) [inline Objects/genericaliasobject.c]:
```suggestion
    // Populate __parameters__ if needed.
```

File: Objects/unionobject.c
Comment (gvanrossum) [inline Objects/unionobject.c]: Same as above.

**Review by gvanrossum — COMMENTED** [review]
LG, I would just update two comments slightly.

File: Objects/unionobject.c
Comment (uriyyo) [reply Objects/unionobject.c]: Changed, thanks for suggestion

File: Lib/test/test_types.py
Comment (uriyyo) [reply Lib/test/test_types.py:675]: Tests have been added, thanks for suggestion

**Comment by Fidget-Spinner** [comment]
Eventually if this PR is merged, we may have to consider backporting this to 3.10. Since it's a somewhat big change, I'm pinging @pablogsal as the 3.10 RM for his awareness.

TLDR: Some uncommon but valid use cases not covered by [PEP 604](https://www.python.org/dev/peps/pep-0604/) were discovered. This PR adds support for them. Can it land in 3.10 beta 4 too?

**Comment by pablogsal** [comment]
> Eventually if this PR is merged, we may have to consider backporting this to 3.10. Since it's a somewhat big change, I'm pinging @pablogsal as the 3.10 RM for his awareness.
> 
> TLDR: Some uncommon but valid use cases not covered by [PEP 604](https://www.python.org/dev/peps/pep-0604/) were discovered in bpo. This PR adds support for them. Can it land in 3.10 beta 4 too?

We are too far in the development cycle to add new features, specially stuff that is touching c code. I do understand that this is something new in 3.10 so we do not have possibility of regressions and I see that this is an important point. I think we are fine to get this in with the following requirements:

* Is merged before the last beta (beta4)
* At least 2 core Devs review and  approve the PR

**Review by gvanrossum — APPROVED** [review]
Thanks!

**Comment by gvanrossum** [comment]
@Fidget-Spinner Can you think of another core dev to review this (not Pablo)?

**Comment by pablogsal** [comment]
> @Fidget-Spinner Can you think of another core dev to review this (not Pablo)?

I can do a review if you don't find anyone else, but currently, I am flooded with PEP 657 stuff so it will probably be in a couple of days

**Comment by Fidget-Spinner** [comment]
> @Fidget-Spinner Can you think of another core dev to review this (not Pablo)?

@serhiy-storchaka , can you review this please?

In the meantime, the Windows 10 buildbot failures seem potentially related so @uriyyo please look into them. I'll take a look too in a day or two.

**Comment by pablogsal** [comment]
> In the meantime, the Windows 10 buildbot failures seem potentially related so @uriyyo please look into them. I'll take a look too in a day or two.

I am restarting them as I am quite sure this is due to corrupted pyc files from a previous run. New builds here:

https://buildbot.python.org/all/#/builders/577/builds/86
https://buildbot.python.org/all/#/builders/593/builds/89

**Comment by Fidget-Spinner** [comment]
> > In the meantime, the Windows 10 buildbot failures seem potentially related so @uriyyo please look into them. I'll take a look too in a day or two.
> 
> I am restarting them as I am quite sure this is due to corrupted pyc files from a previous run. New builds here:
> 
> https://buildbot.python.org/all/#/builders/577/builds/86
> https://buildbot.python.org/all/#/builders/593/builds/89

Oh you're right. Sorry @uriyyo if I led you on a wild goose chase . I suspect *maybe* someone forgot to bump the magic number for Windows. It looks like this PR *may* be the culprit https://github.com/python/cpython/pull/26984.

**Comment by Fidget-Spinner** [comment]
After spending a few hours investigating this. I think our best bet is to merge latest main into this PR and try running the buildbots again. I have no clue why those 2 buildbots fail and any help would be much appreciated.
- They are 2 different workers. I find it unlikely that 2 machines could both have corrupted/out of date pyc files coincidentally.
- I've tried building this locally with same build flags as the buildbots, but I can't reproduce it.
- I noticed the ``buildbot/clean.bat`` script fails on both. But they've been like this for at least a month and I suspect it's because there are no leftover pyc files to clean. Maybe the buildbots already clean it up themselves?

**Comment by gvanrossum** [comment]
Yeah, merge latest main is definitely a good first step.

**Comment by uriyyo** [comment]
@gvanrossum @Fidget-Spinner I have merged latest main into PR branch.

And I have a question should we fix more cases in the scope of this PR?
1. `types.Union` is not serializable and can no be used with `pickle`.
2. `types.Union` can not be used with `typing.Annotated`.
3. `typing._collect_type_vars` currently do not collect params from the `types.Union`.
```py
>>> typing.List[list[T] | float].__parameters__
()
```

**Comment by gvanrossum** [comment]
Tests pass now, I propose to merge now and do another PR for the other issues.

We still need another core dev approval, but I'm merging now. ("Merged before b4" was one of the requirements, "two core devs" is the other.)

**Comment by uriyyo** [comment]
Great, I will open another PR to cover cases that I mentioned.

**Comment by serhiy-storchaka** [comment]
There are several bugs (bpo-44633, bpo-44653), but I am going to fix them soon.

I am +1 for backporting this feature to 3.10.

**Comment by uriyyo** [comment]
@serhiy-storchaka Do you need help with it? There is also issue with `types.Union` serialisation that I am going to fix.

**Comment by pablogsal** [comment]
> There are several bugs ([bpo-44633](https://bugs.python.org/issue44633), [bpo-44653](https://bugs.python.org/issue44653)), but I am going to fix them soon.
> 
> I am +1 for backporting this feature to 3.10.

Please, see my other comment: https://github.com/python/cpython/pull/27207#issuecomment-881913158

**Pull request merged**
